kubernetes-client/src/Kubernetes/Client/Auth/OIDC.hs

```diff
@@ -0,0 +1,182 @@
+                    >>= either (throwM . OIDCOAuthException) pure
+      case OAuth.idToken oauthToken of
+        Nothing -> throwM $ OIDCGetTokenException "token response did not contain an id_token, either the scope \"openid\" wasn't requested upon login, or the provider doesn't support id_tokens as part of the refresh response."
+        Just (IdToken t) -> do
+          _ <- atomically $ writeTVar idTokenTVar (Just t)
+          return t
+
+fetchTokenEndpoint :: Manager -> OIDCAuth -> IO Text
+fetchTokenEndpoint mgr OIDCAuth{..} = do
+  discover issuerURL mgr
+    & (fmap configuration)
+    & (fmap tokenEndpoint)
+
+{-
+   Detects if auth-provier name is oidc, if it is configures the 'KubernetesClientConfig' with OIDCAuth 'AuthMethod'.
+   Does not use cache, consider using 'cachedOIDCAuth'.
+-}
+oidcAuth :: DetectAuth
+oidcAuth AuthInfo{authProvider = Just(AuthProviderConfig "oidc" (Just cfg))} (tls, kubecfg)
+  = Just
+    $ parseOIDCAuthInfo cfg
+    >>= either (throwM . OIDCAuthParsingException) (\oidc -> pure (tls, addAuthMethod kubecfg oidc))
+oidcAuth _ _ = Nothing
+
+-- TODO: Consider doing this whole function atomically, as two threads may miss the cache simultaneously
+{-
+   Detects if auth-provier name is oidc, if it is configures the 'KubernetesClientConfig' with OIDCAuth 'AuthMethod'.
+   First looks for Auth information to be present in 'OIDCCache'. If found returns that, otherwise creates new Auth information and persists it in cache.
+-}
+cachedOIDCAuth :: OIDCCache -> DetectAuth
+cachedOIDCAuth cache AuthInfo{authProvider = Just(AuthProviderConfig "oidc" (Just cfg))} (tls, kubecfg) = Just $ do
+  m <- readTVarIO cache
+  o <- case findInCache m cfg of
+    Left e -> throwM $ OIDCAuthParsingException e
+    Right (Just o) -> return o
+    Right Nothing -> do
+      o@(OIDCAuth{..}) <- parseOIDCAuthInfo cfg
+                          >>= either (throwM . OIDCAuthParsingException) pure
+      let newCache = Map.insert (issuerURL, clientID) o m
+      _ <- atomically $ swapTVar cache newCache
+      return o
+  pure (tls, addAuthMethod kubecfg o)
+cachedOIDCAuth _ _ _ = Nothing
+
+findInCache :: Map (Text, Text) a -> Map Text Text -> Either String (Maybe a)
+findInCache cache cfg = do
+  issuerURL <- lookupEither cfg "idp-issuer-url"
+  clientID <- lookupEither cfg "client-id"
+  return $ Map.lookup (issuerURL, clientID) cache
+
+parseOIDCAuthInfo :: Map Text Text -> IO (Either String OIDCAuth)
+parseOIDCAuthInfo m = do
+  eitherTLSParams <- parseCA m
+  idTokenTVar <- atomically $ newTVar $ Map.lookup "id-token" m
+  refreshTokenTVar <- atomically $ newTVar $ Map.lookup "refresh-token" m
+  return $ do
+    tlsParams <- eitherTLSParams
+    issuerURL <- lookupEither m "idp-issuer-url"
+    clientID <- lookupEither m "client-id"
+    clientSecret <- lookupEither m "client-secret"
+    return OIDCAuth{..}
+
+parseCA :: Map Text Text -> IO (Either String TLS.ClientParams)
+parseCA m = do
```

Comment:
There are a lot of single letter variable names (e.g. `m :: Map Text Text`) that make it hard to piece together what some of these things are.

It might be helpful for future maintainers if either `Map Text Text` were under a `newtype` that made it descriptive, or if `m` was named such that it's more descriptive.

Reply:
I decided to give the variables more understandable name and in general replaced most of the single letter variables with better names. Hopefully it is easier to follow.